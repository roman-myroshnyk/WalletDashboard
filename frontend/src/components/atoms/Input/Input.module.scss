@import '@/styles/colors.scss';

.tag {
  display: flex;
  flex: 1;
  box-sizing: border-box;
  width: 100%;
  height: auto;
  margin: 0;
  padding: 15px;
  overflow: show;

  color: $gray;
  font-weight: 500;

  font-size: 18px;
  font-family: Epilogue, sans-serif;
  font-style: normal;
  line-height: 20px;

  direction: ltr;
  text-align: left;
  text-decoration: none;
  text-overflow: unset;

  background-color: $lightBlack;
  border: none;
  border-radius: 8px;
  outline: none;

  appearance: none;
  resize: none;
  pointer-events: auto;
  -webkit-tap-highlight-color: transparent;
  caret-color: $blue;
}

.tag[type='search']::-webkit-search-cancel-button {
  width: 16px;
  height: 16px;

  background: url('/icons/cancel.svg') no-repeat 50% 50%;
  background-size: contain;
  opacity: 0;

  appearance: none;
  pointer-events: none;
}

.tag[type='search']:focus::-webkit-search-cancel-button {
  opacity: 1;

  pointer-events: all;
}

.tag:disabled {
  flex: unset;

  background-color: transparent !important;
}

.tagTruncate {
  text-overflow: ellipsis;
}

.tag:focus {
  box-shadow: inset 0 0 0 2px $blue !important;
}

.tag::placeholder {
  color: $silverGay !important;
}
