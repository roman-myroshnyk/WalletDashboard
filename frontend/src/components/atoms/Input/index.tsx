/* eslint-disable react/prop-types */

import {
  FC,
  useCallback,
  useRef,
  useMemo,
  InputHTMLAttributes,
  HTMLAttributes,
  ChangeEvent,
  LegacyRef,
  CSSProperties,
} from 'react';
import { useControlledState } from '@/hooks/useControledState';
import styles from './Input.module.scss';

export interface IProps {
    value?: string | number | readonly string[];
    id?:string;
    placeholder?: string;
    onSubmit?: () => void;
    onChange?: (value: string) => void;
    onKeyDown?: (e:KeyboardEvent) => void;
    disabled?: boolean;
    multiLine?: boolean;
    truncate?: boolean;
    blurOnSubmit?: boolean;
    inputType?: InputHTMLAttributes<HTMLInputElement>['type'];
    keyboard?: HTMLAttributes<HTMLInputElement>['inputMode'];
    numberStep?:number;
    numberMin?: number;
    textAlign?: CSSProperties['textAlign'];
}

const Input: FC<IProps> = ({
  value,
  id,
  placeholder,
  onSubmit,
  onChange,
  onKeyDown,
  disabled,
  multiLine,
  truncate,
  blurOnSubmit,
  inputType,
  keyboard,
  numberMin,
  numberStep,
  textAlign,
}) => {
  const [inputValue, setValue] = useControlledState<string | number | readonly string[]>(value);

  const inputEle = useRef<HTMLInputElement | HTMLTextAreaElement>();

  const Tag = useMemo(
    () => (multiLine ? 'textarea' : 'input'),
    [multiLine],
  );

  const handleChange = useCallback(
    (event: ChangeEvent) => {
      const element = multiLine
        ? (event.nativeEvent.target as HTMLTextAreaElement)
        : (event.nativeEvent.target as HTMLInputElement);
      const { value } = element;

      setValue(value);
      if (onChange) onChange(value);
    },
    [multiLine, setValue, onChange],
  );
  const handleKeyDown = useCallback((e:unknown) => {
    if (onKeyDown) onKeyDown(e as KeyboardEvent);
    if ((e as KeyboardEvent).key === 'Enter') {
      if (blurOnSubmit && inputEle.current) { inputEle.current.blur(); }
      if (onSubmit) onSubmit();
    }
  }, [blurOnSubmit, onKeyDown, onSubmit]);
  return (
    <Tag
      style={{
        textAlign,
      }}
      onChange={handleChange}
      ref={inputEle as LegacyRef<HTMLTextAreaElement> & LegacyRef<HTMLInputElement>}
      value={inputValue || ''}
      id={id}
      placeholder={placeholder}
      type={inputType}
      step={numberStep}
      min={numberMin}
      inputMode={keyboard}
      onKeyDown={handleKeyDown}
      disabled={disabled}
      className={`${styles.tag} ${truncate && styles.tagTruncate}`}
      rows={1}
    />
  );
};

export default Input;

Input.defaultProps = {
  value: '',
  id: '',
  placeholder: 'Type somethingâ€¦',
  onSubmit: undefined,
  onChange: undefined,
  onKeyDown: undefined,
  disabled: false,
  multiLine: false,
  truncate: false,
  blurOnSubmit: true,
  inputType: 'text',
  keyboard: 'text',
  numberStep: 0.5,
  numberMin: 1,
  textAlign: 'left',
};
