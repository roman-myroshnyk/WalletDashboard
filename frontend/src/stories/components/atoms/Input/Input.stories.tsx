import { Meta, Story } from '@storybook/react/types-6-0';
import Input, { IProps } from '@/atoms/Input';

export default {
  title: 'atoms/Input',
  component: Input,
  argTypes: {},
} as Meta;

const Template:Story<IProps> = (args) => (
  <Input {...args} />
);

export const Default = Template.bind({});

Default.args = {
};
