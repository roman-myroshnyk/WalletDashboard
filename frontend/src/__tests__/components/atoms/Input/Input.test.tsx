import '.../../../mocks/framer-motion';
import { fireEvent, render } from '@testing-library/react';
import Input from '@/atoms/Input';

describe('atoms/Input', () => {
  it('renders Input with props.value', () => {
    const { container } = render(
      <Input value="I am value" />,
    );
    expect(container).toMatchSnapshot();
  });
  it('renders Input with props.id', () => {
    const { container } = render(
      <Input id="InputId" />,
    );
    expect(container).toMatchSnapshot();
  });
  it('renders Input with props.placeholder', () => {
    const { container } = render(
      <Input placeholder="I am placeholder" />,
    );
    expect(container).toMatchSnapshot();
  });
  it('rendres Input with props.multiLine', () => {
    const { container } = render(
      <Input multiLine />,
    );
    expect(container).toMatchSnapshot();
  });
  it('renders Input with porps.truncate', () => {
    const { container } = render(
      <Input truncate />,
    );
    expect(container).toMatchSnapshot();
  });
  it('renders Input with props.inputType=number', () => {
    const { container } = render(
      <Input inputType="number" />,
    );
    expect(container).toMatchSnapshot();
  });
  it('renders Input with props.keyboard=search', () => {
    const { container } = render(
      <Input keyboard="search" />,
    );
    expect(container).toMatchSnapshot();
  });
  it('renders Input with props.keyboard=numeric', () => {
    const { container } = render(
      <Input keyboard="numeric" />,
    );
    expect(container).toMatchSnapshot();
  });
  it('renders Input with props.numberStep', () => {
    const { container } = render(
      <Input numberStep={3} />,
    );
    expect(container).toMatchSnapshot();
  });
  it('renders Input with props.numberMin', () => {
    const { container } = render(
      <Input numberMin={3} />,
    );
    expect(container).toMatchSnapshot();
  });
  it('renders Input with props.textAlign', () => {
    const { container } = render(
      <Input textAlign="right" />,
    );
    expect(container).toMatchSnapshot();
  });
  it('Input accepts submit event', () => {
    const handleSubmit = jest.fn();
    const { getByPlaceholderText } = render(
      <Input placeholder="input" disabled={false} onSubmit={handleSubmit} />,
    );
    fireEvent.keyDown(getByPlaceholderText('input'), {
      key: 'Enter',
    });
    expect(handleSubmit).toBeCalledTimes(1);
  });
  it('Input accepts change event', () => {
    const handleChange = jest.fn();
    const { getByPlaceholderText } = render(
      <Input placeholder="input" disabled={false} onChange={handleChange} />,
    );
    fireEvent.change(getByPlaceholderText('input'), {
      target: {
        value: 'a',
      },
    });
    expect(handleChange).toBeCalledTimes(1);
  });
  it('Input accepts keydown event', () => {
    const handleKeydown = jest.fn();
    const { getByPlaceholderText } = render(
      <Input placeholder="input" disabled={false} onKeyDown={handleKeydown} />,
    );
    fireEvent.keyDown(getByPlaceholderText('input'), {
      key: 'Enter',
    });
    expect(handleKeydown).toBeCalledTimes(1);
  });
});
