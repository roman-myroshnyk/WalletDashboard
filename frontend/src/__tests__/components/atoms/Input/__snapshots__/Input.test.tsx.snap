// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atoms/Input renders Input with porps.truncate 1`] = `
<div>
  <input
    class="tag tagTruncate"
    data-testid="input"
    id=""
    inputmode="text"
    min="1"
    placeholder="Type something…"
    rows="1"
    step="0.5"
    style="text-align: left;"
    type="text"
    value=""
  />
</div>
`;

exports[`atoms/Input renders Input with props.id 1`] = `
<div>
  <input
    class="tag false"
    data-testid="input"
    id="InputId"
    inputmode="text"
    min="1"
    placeholder="Type something…"
    rows="1"
    step="0.5"
    style="text-align: left;"
    type="text"
    value=""
  />
</div>
`;

exports[`atoms/Input renders Input with props.inputType=number 1`] = `
<div>
  <input
    class="tag false"
    data-testid="input"
    id=""
    inputmode="text"
    min="1"
    placeholder="Type something…"
    rows="1"
    step="0.5"
    style="text-align: left;"
    type="number"
    value=""
  />
</div>
`;

exports[`atoms/Input renders Input with props.keyboard=numeric 1`] = `
<div>
  <input
    class="tag false"
    data-testid="input"
    id=""
    inputmode="numeric"
    min="1"
    placeholder="Type something…"
    rows="1"
    step="0.5"
    style="text-align: left;"
    type="text"
    value=""
  />
</div>
`;

exports[`atoms/Input renders Input with props.keyboard=search 1`] = `
<div>
  <input
    class="tag false"
    data-testid="input"
    id=""
    inputmode="search"
    min="1"
    placeholder="Type something…"
    rows="1"
    step="0.5"
    style="text-align: left;"
    type="text"
    value=""
  />
</div>
`;

exports[`atoms/Input renders Input with props.numberMin 1`] = `
<div>
  <input
    class="tag false"
    data-testid="input"
    id=""
    inputmode="text"
    min="3"
    placeholder="Type something…"
    rows="1"
    step="0.5"
    style="text-align: left;"
    type="text"
    value=""
  />
</div>
`;

exports[`atoms/Input renders Input with props.numberStep 1`] = `
<div>
  <input
    class="tag false"
    data-testid="input"
    id=""
    inputmode="text"
    min="1"
    placeholder="Type something…"
    rows="1"
    step="3"
    style="text-align: left;"
    type="text"
    value=""
  />
</div>
`;

exports[`atoms/Input renders Input with props.placeholder 1`] = `
<div>
  <input
    class="tag false"
    data-testid="input"
    id=""
    inputmode="text"
    min="1"
    placeholder="I am placeholder"
    rows="1"
    step="0.5"
    style="text-align: left;"
    type="text"
    value=""
  />
</div>
`;

exports[`atoms/Input renders Input with props.textAlign 1`] = `
<div>
  <input
    class="tag false"
    data-testid="input"
    id=""
    inputmode="text"
    min="1"
    placeholder="Type something…"
    rows="1"
    step="0.5"
    style="text-align: right;"
    type="text"
    value=""
  />
</div>
`;

exports[`atoms/Input renders Input with props.value 1`] = `
<div>
  <input
    class="tag false"
    data-testid="input"
    id=""
    inputmode="text"
    min="1"
    placeholder="Type something…"
    rows="1"
    step="0.5"
    style="text-align: left;"
    type="text"
    value="I am value"
  />
</div>
`;

exports[`atoms/Input rendres Input with props.multiLine 1`] = `
<div>
  <textarea
    class="tag false"
    data-testid="input"
    id=""
    inputmode="text"
    min="1"
    placeholder="Type something…"
    rows="1"
    step="0.5"
    style="text-align: left;"
    type="text"
  />
</div>
`;
